bpf: monitor: hubble: wireguard: add TRACE_{FROM/TO}_CRYPTO

smagnani96 opened the pull request:
1. `bpf: monitor: hubble: add TRACE_{FROM,TO}_CRYPTO observation point`: add needed structure to support these new trace points, metrics, and hubble/monitor parsing logic for these new types of events.
2. `bpf: wireguard: add trace notification in cil_to_wireguard`: enable the TRACE_TO_CRYPTO from (1) in cil_to_wireguard (cilium_wg0 egress).
3. `bpf: wireguard: test: add wireguard egress hook metrics check`: add new bpf test file for metrics in (3).
4. `bpf: wireguard: TRACE_FROM_CRYPTO in bpf_host`: enable TRACE_FROM_CRYPTO from (1) in cil_from_netdev (cilium_wg0 ingress).
5. `bpf: wireguard: test: add wireguard ingress hook metrics check`: extend existing bpf tests for (5). Will be reverted and placed the logic in (4) when moving to a dedicated Wireguard program cil_from_wireguard.
6. `bpf: nodeport: trace with src_id when available`: enable passing the source identity to `handle_nat_fwd()` for further trace/drop notification.

An example of output with both ingress/egress programs attached on Wireguard (native routing, KPR, nodeEncryption):

```bash
<- endpoint 1418 flow 0x0 , identity 53498->unknown state unknown ifindex 0 orig-ip 0.0.0.0: 10.244.3.84 -> 10.244.1.28 EchoRequest
-> network flow 0x0 , identity 53498->30469 state new ifindex eth0 orig-ip 0.0.0.0: 10.244.3.84 -> 10.244.1.28 EchoRequest
-> crypto flow 0x0 , identity 53498->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 40:00:3f:01:53:89 -> 45:00:00:54:cd:c8 UnknownEthernetType
-> network encrypted  flow 0x0 , identity unknown->remote-node state unknown ifindex eth0 orig-ip 0.0.0.0: 172.18.0.2:51871 -> 172.18.0.4:51871 udp
<- network encrypted  flow 0x0 , identity remote-node->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.4:51871 -> 172.18.0.2:51871 udp
<- crypto flow 0x0 , identity 30469->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 00:00:3f:01:2b:bb -> 45:00:00:54:35:97 UnknownEthernetType
-> endpoint 1418 flow 0x0 , identity 30469->53498 state reply ifindex lxcc04fd937fd43 orig-ip 10.244.1.28: 10.244.1.28 -> 10.244.3.84 EchoReply
```

```bash
~ cilium bpf metrics list
REASON                   DIRECTION   PACKETS   BYTES    LINE   FILE
...
Interface Decrypted      INGRESS     212       80412    1199   bpf_host.c
Interface Decrypting     INGRESS     160       58754    1199   bpf_host.c # <-- from cilium_wg0, will be cilium_from_wireguard in future
Interface Decrypting     INGRESS     42        3655     1156   bpf_host.c
Interface Encrypting     EGRESS      46        3904     267    nodeport_egress.h
Interface Encrypting     EGRESS      47        2941     601    nodeport_egress.h
```

**NOTE**: I opened https://github.com/cilium/cilium/issues/37095 while working on this PR. As you can see from above tracing output, packets being traced from an L3 device (ex. Wireguard) are erroneously decoded. The problem has been identified and discussed in the linked issue, and will be tackled on a separate PR.

Fixes: #35280

```release-note
Add TRACE_{FROM/TO}_CRYPTO observation point and bpf metrics for packets forwarded-to/received-from Wireguard.
```

smagnani96 committed:
bpf: monitor: hubble: add TRACE_{FROM_TO}_CRYPTO observation point

This commit introduces the trace TRACE_{FROM_TO}_CRYPTO observation point
to indicate whether a packet is coming/going through the encryption/decryption
path. This can turn useful for both WireGuard and IPSec, as showed in
subsequent commits.
These values differentiate from the current one as they will be used to
update different types of metrics REASON_{EN,DE}CRYPT that will refers to
different files/lines. This provides more information concerning the flow
of the packet when using `cilium-dbg bpf metrics list`,
`cilium-dbg monitor`, and `hubble observe.
The old TRACE_REASON_ENCRYPTED (along with the REASON_DECRYPT metric) is
still valid and refers to observed encrypted packets prior/after the
crypto process. This patch wants to provide additional metrics and does
not aim to replace those already existing.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

jschwinger233 commented on bpf/bpf_overlay.c:
```diff
@@ -659,6 +659,12 @@ int cil_from_overlay(struct __ctx_buff *ctx)
+#ifdef ENABLE_WIREGUARD
+	if ((ctx->mark & MARK_MAGIC_WG_DECRYPTED) == MARK_MAGIC_WG_DECRYPTED)
+		send_trace_notify(ctx, TRACE_FROM_CRYPTO_DEV, src_sec_identity, UNKNOWN_ID,
+				  TRACE_EP_ID_UNKNOWN, WG_IFINDEX, TRACE_REASON_ENCRYPTED, 0);
```
Using `TRACE_FROM_OVERLAY` seems a bit more reasonable to me. Maybe `TRACE_FROM_CRYPTO_DEV` can be used in the from-wireguard. Similar to `TRACE_TO_CRYPTO_DEV`, which can be used in to-wireguard.

smagnani96 replied:
Agree. This will be addressed in an upcoming work.

jschwinger233 reviewed (commented):
Adding `from-wireguard` to trace decryption events makes sense to me. But I think we can use `TRACE_{FROM,TO}_CRYPTO_DEV` in `from-wireguard` and `to-wireguard` instead of anywhere in bpf_host. Those two hooks see the plain-text packets which should be easy to retrieve src_id for trace.

If the above point also makes sense to you, then maybe we don't have to add new mark `MARK_MAGIC_WG_DECRYPTED`.

smagnani96 committed:
bpf: wireguard: add trace notification in cil_to_wireguard

This commit adds send_trace_notify into the cil_to_wireguard.
With no monitor aggregations, the notification is logged using the new
TRACE_TO_CRYPTO observation point.
With monitor aggregation, the trace event will be displayed according to
the retrieved `monitor` value. In any case, the code updated the respective
bpf metric, and it shows up with `cilium bpf metrics list`, meaning that
packets are going through the WireGuard encryption hook.

Note that we attach the `cil_to_wireguard` program only given a set of
condition. In all the other cases, no metrics updates nor trace events
are expected from this program. Always-injecting wg-related program will
be considered for future enhancements.

An example of a trace event would look as follows:

```bash
-> crypto flow 0x0 , identity unknown->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 40:00:3f:01:48:2f -> 45:00:00:54:d9:f3 UnknownEthernetType
```

The bpf metrics map would contain also the following entry:

```bash
REASON                   DIRECTION   PACKETS   BYTES       LINE   FILE
...
Interface Encrypting     EGRESS      1218      103428      57    bpf_wireguard.c
...
```

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: wireguard: TRACE_FROM_CRYPTO in bpf_host

This commit modifies the tracing observation point in cil_from_netdev
in case the program is injected into the Wireguard device. We do that
to attach `cil_from_netdev` to the ingress hook of the wg device given
a set of conditions. In that case, the tracing system would still use
TRACE_FROM_NETWORK as observation point: let's update it with the new
TRACE_FROM_CRYPTO one (a from-network even would already be traced from
the native device).

This commit also adds the minimal logic in `trace.h` to handle
`{FROM,TO}_CRYPTO` events either coming from bpf_wireguard.c or bpf_host.c.
Using these observation points outside the expected code path
(ENABLE_WIREGUARD && (IS_BPF_HOST || IS_BPF_WIREGUARD)) would result in
a build bug.

An example of a trace event would look as follows:

```bash
<- crypto flow 0x0 , identity unknown->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 40:00:3f:01:48:2f -> 45:00:00:54:d9:f3 UnknownEthernetType
```

The bpf metrics map would contain also the following entry:

```bash
REASON                   DIRECTION   PACKETS   BYTES       LINE   FILE
...
Interface Decrypting     INGRESS     7303      470892      1199   bpf_host.c
...
```

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: wireguard: test: add wireguard egress hook metrics check

This commit introduces a new bpf test file to make sure that `cil_to_wireguard`
correctly updates the metrics map with REASON_ENCRYPTED.
Here we do not explicitly test the whole egress hook logic (red-DNAT in
handle_nat_fwd) as already covered in other test files. For this reason,
the remaining aspect to cover is the correct metric update.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: wireguard: test: add wireguard ingress hook metrics check

This commit modifies the current tc_nodeport_l3_dev logic to be executed
from the wireguard device (IS_BPF_WIREGUARD and ENABLE_WIREGUARD).
Unfortunately, our Wireguard ingress hook is represented by `cil_from_netdev`,
therefore here we cannot add such a check in the apposite wireguard_metrics.c
file. Once we'll move to a custom `cil_from_wireguard` then we can simply
test this logic there and restore the current file by decoupling-it from
IS_BPF_WIREGUARD.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

jschwinger233 commented on bpf/bpf_overlay.c:
```diff
@@ -687,7 +687,7 @@ int cil_from_overlay(struct __ctx_buff *ctx)
 		 */
 		if (decrypted)
-			obs_point = TRACE_FROM_STACK;
+			obs_point = TRACE_FROM_CRYPTO;
```
Just a note for 1.18: decrypted packets (previous ESP) will no longer get to from_overlay because of VinE.

smagnani96 replied:
Yep, thank you very much for the note. Dropping the IPSec-related part from this PR and moving to a future one.

jschwinger233 reviewed (commented):
So the new TRACE_FROM|TO_CRYPTO ob points are tracing plain-text traffic right after decryption|before encryption. Previously we don't have these ob points; TRACE_REASON_ENCRYPTED is used to report ESP traffic before decryption, but we don't know the inner packet.

The reasoning makes sense to me, but maybe we want to separate the ipsec changes to a new PR? Mixing wg and ipsec in a big PR requires extra context and carefulness.

jschwinger233 commented on bpf/lib/nodeport_egress.h:
```diff
@@ -105,7 +105,7 @@ int tail_handle_snat_fwd_ipv6(struct __ctx_buff *ctx)
 #ifdef IS_BPF_OVERLAY
 	obs_point = TRACE_TO_OVERLAY;
-#elif defined(IS_BPF_WIREGUARD)
+#elif defined(IS_BPF_WIREGUARD) && IS_BPF_WIREGUARD
```
Maybe I'm wrong: `defined(IS_BPF_WIREGUARD) && IS_BPF_WIREGUARD` is always false during clang compiling, so `obs_point = TRACE_TO_CRYPTO` is always compiled out. The further loader injection `consts["IS_BPF_WIREGUARD"] = 1` is too late to bring back the `obs_point = TRACE_TO_CRYPTO` statement.

Did you test and confirm that we can see TRACE_TO_CRYPTO from tail_handle_snat_fwd ? If you tested, then I must be wrong.

smagnani96 replied:
Correct, but this is also what I was expecting:
1. when called from `bpf_wireguard.c`, I want to show that the packet was TRACE_TO_CRYPTO
2. when called from `bpf_host.c`, I want to keep it TRACE_TO_NETWORK. Since IS_BPF_WIREGUARD would still be defined, I want to make sure that condition doesn't pass. Therefore, checking for true, but that would always be false at compile-time in bpf_host.

Does it make sense?

jschwinger233 replied:
I thought what you want is to use TRACE_TO_CRYPTO when tail_handle_snat_fwd_ipv6 is called from to_netdev@cilium_wg0? 

If you want to keep TRACE_TO_NETWORK, why do we add:
```c
DEFINE_BOOL(IS_BPF_WIREGUARD, false);
#define IS_BPF_WIREGUARD fetch_bool(IS_BPF_WIREGUARD)
```
to_netdev@cilium_wg0 doesn't have IS_BPF_WIREGUARD in the first place, we already have "when called from bpf_host.c, keep it TRACE_TO_NETWORK".

smagnani96 replied:
I need that because I'd like TRACE_FROM_CRYPTO in `cil_from_netdev` when attached to Wireguard (ingress).

jschwinger233 replied:
okay I was confused of the direction: only from_netdev will be attached to cilium_wg0. Ignore the nonsense I said above. I believe your code is right, logically.

Instead of defining a `IS_BPF_WIREGUARD` in bpf_host, how about doing it this way:
```diff
- enum trace_point obs_point = IS_BPF_WIREGUARD ? TRACE_FROM_CRYPTO : TRACE_FROM_NETWORK;
+ enum trace_point obs_point = THIS_INTERFACE_IFINDEX == WG_IFINDEX ? TRACE_FROM_CRYPTO : TRACE_FROM_NETWORK;
```

I feel it's easier?

smagnani96 replied:
Many thanks for the suggestion, let me try this out and I'll back to this soon.
What you're proposing feels way more intuitive: having IS_BPF_WIREGUARD from bpf_host would cause headaches, especially considering that it would be a temporary solution until we move to a new `cil_from_wireguard`.
So yes, I very like `THIS_INTERFACE_IFINDEX == WG_IFINDEX`.

jschwinger233 reviewed (commented):
One newbie question regarding `#if defined(IS_BPF_WIREGUARD) && IS_BPF_WIREGUARD`. If you already tested with positive results, don't be bothered. :thanks:

smagnani96 committed:
bpf: nodeport: trace with src_id when available

This commit adapts the `handle_nat_fwd()` function to accept the src_id
that can be used for subsequent trace/drop notification. The caller of
that function has already retrieved the source identity, so let's reuse
it via the skb meta. Such meta is always cleared via `ctx_load_and_clear_meta()`.

In the `nat_fwd` functions like `tail_handle_nat_fwd_ipv{4,6}()`, we first
retrieve value w/o clearing via `ctx_load_meta`, and then we always invoke
their internal handler functions `handle_nat_fwd_ipv{4,6}`, which properly
clears our the meta value.
On the other hand, in the `snat_fwd` functions, we directly clear the
value upon its retrieval, as their handler functions don't need to use it.

An example of a trace before and after this patch could be as follows:

```bash
~ Before
-> crypto flow 0x0 , identity unknown->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 40:00:3f:01:48:2f -> 45:00:00:54:d9:f3 UnknownEthernetType
~ After
-> crypto flow 0x0 , identity 21567->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 40:00:3f:01:48:2f -> 45:00:00:54:d9:f3 UnknownEthernetType
```

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

kaworu commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -534,6 +534,22 @@ func decodeTrafficDirection(srcEP uint32, dn *monitor.DropNotify, tn *monitor.Tr
+			monitorAPI.TraceFromCrypto,
+			monitorAPI.TraceFromNetwork:
+			return pb.TrafficDirection_INGRESS
+		}
```
guessing ingress/egress based on the observation point make sense to me intuitively, I wonder why we didn't do it sooner cc @gandro who has more context around this codepath than I do.

Also question: is `TRACE_TO_LXC → EGRESS` and `TRACE_TO_LXC → INGRESS` correct here? Isn't it the other way around?

gandro replied:
> guessing ingress/egress based on the observation point make sense to me intuitively, I wonder why we didn't do it sooner cc @gandro who has more context around this codepath than I do.

We did consider this, but we ended up not doing it because I'm afraid it's not correct. The traffic direction field is supposed to show the direction of the **connection** the packet belongs to, not the direction of the packet itself. 

So a reply packet (egressing the node) for an ingress connection should still be classified as ingress. I'd rather leave the traffic direction field empty here.

smagnani96 replied:
> The traffic direction field is supposed to show the direction of the connection the packet belongs to, not the direction of the packet itself.

That explains a lot, thanks for clarifying it :pray: 
Then, also for  `TRACE_{FROM,TO}_CRYPTO` events, I'd let the parser try inferring it from the trace reason and/or the other info known, removing this portion of code (and commits) from the PR.

julianwiedmann commented on bpf/lib/nodeport_egress.h:
```diff
@@ -105,6 +105,8 @@ int tail_handle_snat_fwd_ipv6(struct __ctx_buff *ctx)
+	obs_point = TRACE_TO_CRYPTO;
 #else
 	obs_point = TRACE_TO_NETWORK;
 #endif
```
I wonder if this should just be a macro that is defined by all programs which include `nodeport_egress.h`?

```
#define NODEPORT_OBS_POINT_EGRESS      TRACE_TO_CRYPTO
```

smagnani96 replied:
> I wonder if this should just be a macro that is defined by all programs which include nodeport_egress.h?

Mmm what about moving the define at the top of `nodeport_egress.h`?
```c
#if defined(IS_BPF_OVERLAY)
#define NODEPORT_OBS_POINT_EGRESS      TRACE_TO_OVERLAY
#elif defined(IS_BPF_WIREGUARD)
#define NODEPORT_OBS_POINT_EGRESS      TRACE_TO_CRYPTO
#else
#define NODEPORT_OBS_POINT_EGRESS      TRACE_TO_NETWORK
#endif
```

julianwiedmann replied:
works for me!

jschwinger233 commented on bpf/lib/nodeport_egress.h:
```diff
@@ -393,6 +403,8 @@ static __always_inline int nodeport_snat_fwd_ipv4(struct __ctx_buff *ctx,
 {
+	__u32 __maybe_unused src_sec_identity = UNKNOWN_ID;
+	__u32 magic = ctx->mark & MARK_MAGIC_HOST_MASK;
 	__u32 cluster_id = ctx_load_and_clear_meta(ctx, CB_CLUSTER_ID_EGRESS);
```
I feel we'd better use the same way as `cluster_id = ctx_load_and_clear_meta(ctx, CB_CLUSTER_ID_EGRESS);`:

```diff
- __u32 magic = ctx->mark & MARK_MAGIC_HOST_MASK;
+ __u32 src_sec_identity = ctx_load_and_clear_meta(ctx, CB_SRC_LABEL);
```

Also, this requires CB_SRC_LABEL to be set from callers, e.g. `handle_nat_fwd` https://github.com/cilium/cilium/blob/v1.17.0/bpf/lib/nodeport_egress.h#L712
```diff
+ ctx_store_meta(ctx, CB_SRC_LABEL, src_id);
```
the `src_id` is already known in the caller of `handle_nat_fwd`, just pass it in and store it in cb.

smagnani96 replied:
Thanks for this observation. I just updated the code following this style and made sure that the outcome was the same. Let me know :smiley:

jschwinger233 reviewed (commented):
(nonblocking, good to have as a followup) Just have a feeling that setting src_sec_id in cb and retrieving it there later is a more common pattern in cilium tailcall. Others LGTM.
